evap/evaluation/auth.py

```diff
@@ -149,3 +152,44 @@ def check_user(user):
         from evap.rewards.tools import can_reward_points_be_used_by
         return can_reward_points_be_used_by(user)
     return user_passes_test(check_user)(view_func)
+
+
+# see https://mozilla-django-oidc.readthedocs.io/en/stable/
+class OpenIDAuthenticationBackend(OIDCAuthenticationBackend):
+    def filter_users_by_claims(self, claims):
+        email = claims.get('email')
+        if not email:
+            return self.UserModel.objects.none()
+
+        user_email = email
+        for original_domain, replaced_domain in settings.INSTITUTION_EMAIL_REPLACEMENTS:
+            if email.endswith(original_domain):
+                user_email = email[:-len(original_domain)] + replaced_domain
+                break
+
+        try:
+            return UserProfile.objects.filter(email=user_email)
+        except UserProfile.DoesNotExist:
+            return self.UserModel.objects.none()
+
+    def create_user(self, claims):
+        user = super(OpenIDAuthenticationBackend, self).create_user(claims)
+
+        user.email = claims.get('email')
```

Comment:
We want to set the username from the part of the email before the "@" here afaik.